#Ansible Module Name -- rds.py
#Test Case : Modify RDS instance with command parameter 
#Ansible Playbook created by - Sachin J
---

- name: Modify RDS instance with command parameter 
  hosts: localhost
  connection: local
  vars_files:
     - ../../../vars/alicloud-creds.yml
     - ../../vars/test-environment.yml
     - ../../vars/general-parameters.yml
  
  tasks:
  
 # Task : 1
   - block:
     - name: Scenario 1 - Create an RDS instance to modify
       rds:
         acs_access_key: '{{ aliyun_acs_access_key }}'
         acs_secret_access_key: '{{ aliyun_acs_secret_access_key }}'
         region: '{{ region }}'
         #zone: '{{ zone }}'
         command: '{{ command }}'
         db_engine: '{{ dbEngineMysql }}'
         engine_version: '{{ engineVersion }}'
         db_instance_storage: '{{ dbInstanceStorage }}'
         db_instance_class: '{{ dbInstanceClass }}'
         instance_net_type: '{{ instanceNetTypeInternet }}'
         security_ip_list: '{{ securityIpList }}'
         pay_type: '{{ payType }}'
      
       register: result_scn1
     - debug: var=result_scn1
    
     rescue:
      - fail:
         msg: "The fail block is added just to handle exception. This block won't be executed ever."
        when: no

   - pause:
       minutes: 3
      
      
  # Task : 2
  
   - block: 
     - name: Scenario 2 - Modify an RDS instance with command parameter as invalid
       rds:
         acs_access_key: '{{ aliyun_acs_access_key }}'
         acs_secret_access_key: '{{ aliyun_acs_secret_access_key }}'
         region: '{{ region }}'
         #zone: '{{ zone }}'
         command: '{{ commandInvalid }}'
         instance_id: '{{ result_scn1.result[0].DBInstanceId }}'
         security_ip_list: '{{ securityIpListModify }}'
         db_instance_storage: '{{ dbInstanceStorage }}'
         pay_type: '{{ payType }}'   
         connection_mode: Performance   
       

       register: result_scn2
     - debug: var=result_scn2
     
     rescue:
      - fail:
         msg: "The fail block is added just to handle exception. This block won't be executed ever."
        when: no 
        
        
   # Task : 3
  
   - block: 
     - name: Scenario 3 - Modify an RDS instance with command parameter as create
       rds:
         acs_access_key: '{{ aliyun_acs_access_key }}'
         acs_secret_access_key: '{{ aliyun_acs_secret_access_key }}'
         region: '{{ region }}'
         #zone: '{{ zone }}'
         command: '{{ command }}'
         instance_id: '{{ result_scn1.result[0].DBInstanceId }}'
         security_ip_list: '{{ securityIpListModify }}'
         db_instance_storage: '{{ dbInstanceStorage }}'
         pay_type: '{{ payType }}'   
         connection_mode: Performance   
       

       register: result_scn3
     - debug: var=result_scn3
     
     rescue:
      - fail:
         msg: "The fail block is added just to handle exception. This block won't be executed ever."
        when: no      
        
   # Task : 4
  
   - block: 
     - name: Scenario 4 - Release the created RDS instance
       rds:
        access_key: '{{ aliyun_acs_access_key }}'
        secret_key: '{{ aliyun_acs_secret_access_key }}'
        command: '{{ commandDelete }}'
        instance_id: '{{ result_scn1.result[0].DBInstanceId }}'
        region: '{{ region }}'
        
       register: result_scn4
     - debug: var=result_scn4
    
     rescue:
      - fail:
         msg: "The fail block is added just to handle exception. This block won't be executed ever."
        when: no  
         
     always: 
      - assert: 
         that: 
          - "'failed' not in result_scn1" 
          - "'failed' in result_scn2" 
          - "'failed' in result_scn3"
          - "'failed' not in result_scn4"  
                    
         msg: 'Test pass FAILED !!!'   
       